moshi/src/main/java/com/squareup/moshi/ClassFactory.java

```diff
@@ -30,25 +29,55 @@
  * @author Jesse Wilson
  */
 abstract class ClassFactory<T> {
+
+  /**
+   * This can't be a {@link LinkedHashTreeMap} since {@link Class}
+   * does not implement {@link Comparable}.
+   */
+  private static Map<Class<?>, ClassFactory<?>> unsafeFactories = new HashMap<>();
```

Comment:
The adapter may not always be fast without caching due to the recursive fallback call to `ClassFactory.get(parameterType).newInstance()`. It's your call on this though, I don't really care too much about the caching since many real-world entities consist of primitives and `String`s only anyway.